rpc: remove deprecated CRPCCommand constructor

maflcko opened the pull request:
Remove the CRPCCommand arguments, now that they are asserted to be equal and thus redundant

### Future work

> Here or follow up, makes sense to also assert type of returned UniValue?

Sure, but let's not get ahead of ourselves. I am going to submit any further works as follow-ups, including:

* Removing all python regex linters on the args, now that RPCMan can be used to generate any output, including the cli.cpp table
* Auto-formatting and sanity checking the RPCExamples with RPCMan
* Checking passed-in json in self-check. Removing redundant checks
* Checking returned json against documentation to avoid regressions or false documentation
* Compile the RPC documentation at compile-time to ensure it doesn't change at runtime and is completely static

### Bugs found

* The assert identified issue #18607 
* The changes itself fixed bug #19250

Commit:
remove CRPCCommand constructor that takes rpcfn_type function pointer

promag reviewed (commented):
Concept ACK.

Here or follow up, makes sense to also assert type of returned UniValue?

MarcoFalke commented:
> Here or follow up, makes sense to also assert type of returned UniValue?

Sure, but let's not get ahead of ourselves. I am going to submit any further works as follow-ups, including:

* Removing the CRPCCommand arguments, now that they are asserted to be equal and thus redundant
* Removing all python regex linters on the args, now that RPCMan can be used to generate any output, including the cli.cpp table
* Auto-formatting and sanity checking the RPCExamples with RPCMan
* Checking passed-in json in self-check. Removing redundant checks
* Checking returned json against documentation to avoid regressions or false documentation

MarcoFalke commented:
Please review the bugfix in #18607 first

MarcoFalke commented:
Rebased

pierreN commented on src/rpc/server.cpp:
```diff
@@ -140,44 +138,54 @@ UniValue help(const JSONRPCRequest& jsonRequest)
-            }.ToString()
+        [&](const RPCMan& self, const JSONRPCRequest& jsonRequest) -> UniValue
+{
+    self.Check(jsonRequest);
+    if (jsonRequest.fHelp || jsonRequest.params.size() > 1)
+        throw std::runtime_error(
+            self.ToString()
```
Since you add a call to `RPCMan::Check`, why then recheck again for `fHelp` flag/`params` size? The if statement could be removed, right?

pierreN reviewed (approved):
Tested ACK fa45fbc

This may be out of scope for this PR, but something that might bother `RPCMan` users a bit is to have to call `Check` every time (as most do now).
I think this issue is reflected in the lambda function signature: to do the actual RPC work, why would you need a reference to `RPCMan`?

I tried to refactor a bit the changes on top of your PR so that it's easier for users to use `RPCMan`: https://github.com/pierreN/bitcoin/commit/b1633b0e449177ca5eea382409e3a3af064fde08  (splits the lambda in 2, add a constructor helper to automatically call `Check`)

IMHO this makes the class a tad easier to use (this might be personal taste though?) and shaves off ~500 lines. 
If you think this refactor is a good idea, please do add it to the PR. Or if you think this is too out of scope, I could do a PR later once this one is merged (or we could just let things as is).

MarcoFalke commented:
@pierreN Thanks for the review. I do like your suggestion to pass in the checks into RPCMan and execute them there. However, I think this can be done nicely in a follow-up and would bloat this pull a bit too much.

MarcoFalke commented:
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

pierreN commented:
> I do like your suggestion to pass in the checks into RPCMan and execute them there. However, I think this can be done nicely in a follow-up and would bloat this pull a bit too much.

Thanks, indeed. I'll send a PR once this one is merged then :smiley:

MarcoFalke commented:
Rebased

MarcoFalke commented:
Rebased

MarcoFalke commented:
Rebased

MarcoFalke commented:
Rebased

MarcoFalke commented:
Edited OP to summarize the motivation, future work, and bugs found by this check. Also, rebased

MarcoFalke commented:
Please review #19250 first, which has been extracted as a separate commit and pull request

MarcoFalke commented:
Rebased to reduce the diff and get rid of the already merged bugfixes.

Not sure why GitHub is hiding the comments (there are only 8 in total), but for reference this has one Concept ACK and one (stale) tested ACK.

MarcoFalke commented:
Rebased due to conflict in adjacent line

jonatack commented:
Concept ACK, will review.

For testing the printed JSON of CAmounts, I propose `assert_decimal_places` in test_framework/util.py in 741d4b94f4f30ca2d7e4886bc98a9faa8d5cc8c2 used for testing bitcoin balances in #19089 and #19092.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -2435,39 +2435,39 @@ void RegisterBlockchainRPCCommands(CRPCTable &t)
 static const CRPCCommand commands[] =
 { //  category              name                      actor (function)         argNames
```
Missed fixing the headline here.

fjahr reviewed (commented):
Concept ACK

Will test when this is rebased and checks pass. Please re-check the commit messages, 97a419abaf673d072e3c5b23af078e403d1c5dc4 and 0bd51da9ecdc0c5f3834f8824d260f6f535ded92 seem to be outdated.

MarcoFalke commented:
Current one to review is #19717 ;)

MarcoFalke commented:
Please review #19849 first.

Commit:
remove dead rpc code

Checking for fHelp, or the size of the args, is dead code because:
* fHelp is always false (src/qt/test/rpcnestedtests.cpp)
* It is already implicitly called by RPCHelpMan::Check
  (src/rpc/mining.cpp, src/rpc/misc.cpp, src/rpc/net.cpp)

MarcoFalke commented:
Please review #19994 first.

MarcoFalke commented:
This is ready for review now

promag reviewed (commented):
On fa19bb2cd8c575593583138a84e6bb3444d6196d you could rename `fHelp` so that if this gets rebased we don't miss new cases.

promag reviewed (commented):
Tested ACK faaf9c58e4aa809019d4ca12747dd47411988e37.

Verified 1st commit claims.

fjahr commented:
tested ACK faaf9c58e4aa809019d4ca12747dd47411988e37

MarcoFalke commented on src/rpc/server.h:
```diff
@@ -85,7 +85,6 @@ void RPCUnsetTimerInterface(RPCTimerInterface *iface);
 void RPCRunLater(const std::string& name, std::function<void()> func, int64_t nSeconds);
 
-typedef UniValue(*rpcfn_type)(const JSONRPCRequest& jsonRequest);
 typedef RPCHelpMan (*RpcMethodFnType)();
```
The def is also used in #20012

ryanofsky reviewed (approved):
Code review ACK faaf9c58e4aa809019d4ca12747dd47411988e37. Two obviously good simplifications.

re: https://github.com/bitcoin/bitcoin/pull/18531#issue-399144769

> Remove the CRPCCommand arguments, now that they are asserted to be equal and thus redundant
> 
> ### Future work
> > Here or follow up, makes sense to also assert type of returned UniValue?
> 
> Sure, but let's not get ahead of ourselves. I am going to submit any further works as follow-ups, including:
> 
> * Removing all python regex linters on the args, now that RPCMan can be used to generate any output, including the cli.cpp table
> * Auto-formatting and sanity checking the RPCExamples with RPCMan
> * Checking passed-in json in self-check. Removing redundant checks
> * Checking returned json against documentation to avoid regressions or false documentation
> * Compile the RPC documentation at compile-time to ensure it doesn't change at runtime and is completely static
> 
> ### Bugs found
> * The assert identified issue #18607
> * The changes itself fixed bug #19250

I don't see this is mentioned in "future work" part of PR description, but it seems like a good followup would be to drop redundant `name_in` and `args_in` arguments to the other `CRPCCommand` constructor:

https://github.com/MarcoFalke/bitcoin-core/blob/faaf9c58e4aa809019d4ca12747dd47411988e37/src/rpc/server.h#L106-L116

since these are just checked for equality and discarded anyway. Or maybe that's what the PR was originally intended to do? The code changes are simple, but the PR description is a little hard to follow

MarcoFalke commented:
> since these are just checked for equality and discarded anyway. Or maybe that's what the PR was originally intended to do? The code changes are simple, but the PR description is a little hard to follow

Indeed. Done in #20012
